File: packages/custos-py/src/custos/proxy.py
"""Transparent MCP stdio proxy.

Reads JSON-RPC messages from stdin, forwards to an upstream MCP server process
over stdio, gates `tools/call` methods through the policy engine, and appends
signed decision records to the ledger.
"""
from __future__ import annotations

import asyncio
import json
import sys
import time
from dataclasses import dataclass
from typing import Optional

from custos.ids import iso_now_ms, new_span_id, new_trace_id
from custos.ledger import Ledger, hash_of_value
from custos.policy import Policy
from custos.record import Actor, Decision, DecisionRecord, PolicyResult, Server


DENY_CODE = -32001


@dataclass
class ProxyConfig:
    upstream_cmd: list[str]
    policy: Policy
    ledger: Ledger
    actor: Actor
    server: Server


async def _read_line(stream: asyncio.StreamReader) -> Optional[bytes]:
    try:
        line = await stream.readline()
    except Exception:
        return None
    if not line:
        return None
    return line


async def run_stdio_proxy(config: ProxyConfig) -> int:
    """Run the proxy until the upstream process or stdin closes."""
    proc = await asyncio.create_subprocess_exec(
        *config.upstream_cmd,
        stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=sys.stderr,
    )
    assert proc.stdin and proc.stdout

    # Bridge stdin (from client) → policy → upstream stdin
    loop = asyncio.get_event_loop()
    reader = asyncio.StreamReader()
    protocol = asyncio.StreamReaderProtocol(reader)
    await loop.connect_read_pipe(lambda: protocol, sys.stdin)

    # Track pending tool calls by JSON-RPC id, so we can hash results
    pending: dict = {}  # id -> {"tool", "args", "trace_id", "span_id", "t0"}

    async def client_to_upstream():
        while True:
            line = await _read_line(reader)
            if line is None:
                break
            try:
                msg = json.loads(line)
            except Exception:
                # Non-JSON: pass through
                proc.stdin.write(line)
                await proc.stdin.drain()
                continue
            if isinstance(msg, dict) and msg.get("method") == "tools/call":
                params = msg.get("params") or {}
                tool = params.get("name", "")
                args = params.get("arguments", {}) or {}
                meta = params.get("_meta") or {}
                trace_id = meta.get("trace_id") or new_trace_id()
                span_id = new_span_id()
                params["_meta"] = {**meta, "trace_id": trace_id, "span_id": span_id}
                ctx = {
                    "tool": tool,
                    "actor": {
                        "id": config.actor.id,
                        "kind": config.actor.kind,
                        "meta": config.actor.meta,
                    },
                    "server": config.server.to_dict(),
                    "args": args,
                    "trace_id": trace_id,
                }
                pd = config.policy.evaluate(ctx)
                if pd.decision != Decision.ALLOW:
                    # Deny: reply to client, do not forward
                    rec = _record(
                        config, tool, args, None, pd.decision, pd.rule_id, pd.reason,
                        latency_ms=0, trace_id=trace_id, span_id=span_id,
                    )
                    config.ledger.append(rec)
                    err_resp = {
                        "jsonrpc": "2.0",
                        "id": msg.get("id"),
                        "error": {
                            "code": DENY_CODE,
                            "message": f"denied by policy: {pd.rule_id or 'default'}",
                            "data": {"reason": pd.reason, "trace_id": trace_id},
                        },
                    }
                    sys.stdout.write(json.dumps(err_resp) + "\n")
                    sys.stdout.flush()
                    continue
                pending[msg.get("id")] = {
                    "tool": tool,
                    "args": args,
                    "trace_id": trace_id,
                    "span_id": span_id,
                    "t0": time.perf_counter(),
                    "rule": pd.rule_id,
                    "reason": pd.reason,
                }
                proc.stdin.write((json.dumps(msg) + "\n").encode("utf-8"))
                await proc.stdin.drain()
            else:
                proc.stdin.write(line)
                await proc.stdin.drain()

    async def upstream_to_client():
        while True:
            line = await proc.stdout.readline()
            if not line:
                break
            try:
                msg = json.loads(line)
            except Exception:
                sys.stdout.buffer.write(line)
                sys.stdout.flush()
                continue
            if isinstance(msg, dict) and msg.get("id") in pending:
                info = pending.pop(msg["id"])
                latency_ms = int((time.perf_counter() - info["t0"]) * 1000)
                result = msg.get("result")
                error = msg.get("error")
                decision = Decision.ALLOW if error is None else Decision.ERROR
                reason = info["reason"] if decision == Decision.ALLOW else f"upstream error: {error}"
                rec = _record(
                    config, info["tool"], info["args"], result,
                    decision, info["rule"], reason,
                    latency_ms=latency_ms, trace_id=info["trace_id"], span_id=info["span_id"],
                )
                config.ledger.append(rec)
            sys.stdout.buffer.write(line)
            sys.stdout.flush()

    await asyncio.gather(client_to_upstream(), upstream_to_client(), return_exceptions=True)
    return await proc.wait()


def _record(
    config: ProxyConfig,
    tool: str,
    args,
    result,
    decision: Decision,
    rule: str,
    reason: str,
    latency_ms: int,
    trace_id: str,
    span_id: str,
) -> DecisionRecord:
    return DecisionRecord(
        v=1,
        seq=0,
        ts=iso_now_ms(),
        trace_id=trace_id,
        span_id=span_id,
        actor=config.actor,
        server=config.server,
        tool=tool,
        args_hash=hash_of_value(args),
        result_hash=hash_of_value(result) if (decision == Decision.ALLOW and result is not None) else "",
        decision=decision,
        policy=PolicyResult(
            engine=config.policy.engine, id=config.policy.id, rule=rule, reason=reason
        ),
        latency_ms=latency_ms,
        prev_hash="",
    )

File: packages/custos-js/src/proxy.ts
import { spawn, ChildProcess } from "node:child_process";
import { createInterface } from "node:readline";

import { isoNowMs, newSpanId, newTraceId } from "./ids.js";
import { hashOfValue, Ledger } from "./ledger.js";
import { Policy } from "./policy.js";
import { Actor, Decision, DecisionRecord, Server, serverToDict } from "./record.js";
import { generateToken } from "./token.js";

const DENY_CODE = -32001;

export interface ProxyConfig {
  upstreamCmd: string[];
  policy: Policy;
  ledger: Ledger;
  actor: Actor;
  server: Server;
}

export async function runStdioProxy(cfg: ProxyConfig): Promise<number> {
  const [cmd, ...args] = cfg.upstreamCmd;
  if (!cmd) throw new Error("upstream cmd required");
  const proc: ChildProcess = spawn(cmd, args, { stdio: ["pipe", "pipe", "inherit"] });

  const pending = new Map<string | number, {
    tool: string; args: any; traceId: string; spanId: string;
    t0: number; rule: string; reason: string;
  }>();

  const clientRl = createInterface({ input: process.stdin });
  const upstreamRl = createInterface({ input: proc.stdout! });

  clientRl.on("line", (line: string) => {
    if (!line.trim()) return;
    let msg: any;
    try { msg = JSON.parse(line); } catch { proc.stdin!.write(line + "\n"); return; }
    if (msg && msg.method === "tools/call") {
      const params = msg.params ?? {};
      const tool = params.name ?? "";
      const args = params.arguments ?? {};
      const meta = params._meta ?? {};
      const traceId = meta.trace_id ?? newTraceId();
      const spanId = newSpanId();
      params._meta = { ...meta, trace_id: traceId, span_id: spanId };
      const ctx = {
        tool,
        actor: { id: cfg.actor.id, kind: cfg.actor.kind, meta: cfg.actor.meta },
        server: serverToDict(cfg.server),
        args,
        trace_id: traceId,
      };
      const pd = cfg.policy.evaluate(ctx);
      if (pd.decision !== "allow") {
        const rec = buildRecord(cfg, tool, args, undefined, pd.decision, pd.ruleId, pd.reason, 0, traceId, spanId);
        cfg.ledger.append(rec);
        const err = {
          jsonrpc: "2.0", id: msg.id,
          error: { code: DENY_CODE, message: `denied by policy: ${pd.ruleId || "default"}`, data: { reason: pd.reason, trace_id: traceId } },
        };
        process.stdout.write(JSON.stringify(err) + "\n");
        return;
      }
      // Attach a signed per-call attestation token to the forwarded
      // call's _meta so a cooperating upstream can prove the call
      // passed through Custos. See WIRE §8.
      try {
        const kp: any = (cfg.ledger as any).kp;
        if (kp && typeof kp.sign === "function") {
          params._meta.custos_token = generateToken(
            kp, traceId, spanId, tool, hashOfValue(args), isoNowMs(),
          );
        }
      } catch {
        // Token generation is best-effort — the ledger still records
        // the decision. Never block the forwarded call over an
        // optional evidence artifact.
      }
      pending.set(msg.id, { tool, args, traceId, spanId, t0: performance.now(), rule: pd.ruleId, reason: pd.reason });
      proc.stdin!.write(JSON.stringify(msg) + "\n");
    } else {
      proc.stdin!.write(line + "\n");
    }
  });

  upstreamRl.on("line", (line: string) => {
    if (!line.trim()) { process.stdout.write("\n"); return; }
    let msg: any;
    try { msg = JSON.parse(line); } catch { process.stdout.write(line + "\n"); return; }
    if (msg && pending.has(msg.id)) {
      const info = pending.get(msg.id)!;
      pending.delete(msg.id);
      const latency = Math.round(performance.now() - info.t0);
      const decision: Decision = msg.error ? "error" : "allow";
      const reason = decision === "allow" ? info.reason : `upstream error: ${JSON.stringify(msg.error)}`;
      const rec = buildRecord(cfg, info.tool, info.args, msg.result, decision, info.rule, reason, latency, info.traceId, info.spanId);
      cfg.ledger.append(rec);
    }
    process.stdout.write(line + "\n");
  });

  return new Promise<number>((resolve) => {
    proc.on("exit", (code) => resolve(code ?? 0));
  });
}

function buildRecord(
  cfg: ProxyConfig, tool: string, args: unknown, result: unknown,
  decision: Decision, rule: string, reason: string,
  latencyMs: number, traceId: string, spanId: string,
): DecisionRecord {
  return {
    v: 1, seq: 0, ts: isoNowMs(),
    trace_id: traceId, span_id: spanId,
    actor: { id: cfg.actor.id, kind: cfg.actor.kind, meta: cfg.actor.meta },
    server: serverToDict(cfg.server),
    tool,
    args_hash: hashOfValue(args),
    result_hash: decision === "allow" && result !== undefined ? hashOfValue(result) : "",
    decision,
    policy: {
      engine: cfg.policy.engine,
      id: cfg.policy.id,
      rule,
      reason,
      ...(cfg.policy.hash ? { hash: cfg.policy.hash } : {}),
    },
    latency_ms: latencyMs,
    prev_hash: "",
    // Stdio proxy is a hard enforcement point: a deny is a JSON-RPC error
    // and the upstream server never sees the forwarded call.
    enforcement: { point: "proxy", effect: "blocked" },
  };
}
